src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -1190,57 +1185,61 @@ private void NotifyFeaturesAboutStateChange(DatabaseRecord record, long index)
+                    _logger.Info($"Starting to process record with index `{index}` (current {LastDatabaseRecordChangeIndex}) for {record.DatabaseName}.");
 
-                    try
-                    {
-                        DatabaseGroupId = record.Topology.DatabaseTopologyIdBase64;
+                try
+                {
+                    DatabaseGroupId = record.Topology.DatabaseTopologyIdBase64;
 
-                        SetUnusedDatabaseIds(record);
-                        InitializeFromDatabaseRecord(record);
-                        IndexStore.HandleDatabaseRecordChange(record, index);
-                        ReplicationLoader?.HandleDatabaseRecordChange(record);
-                        EtlLoader?.HandleDatabaseRecordChange(record);
-                        SubscriptionStorage?.HandleDatabaseRecordChange(record);
+                    SetUnusedDatabaseIds(record);
+                    InitializeFromDatabaseRecord(record);
+                    IndexStore.HandleDatabaseRecordChange(record, index);
+                    ReplicationLoader?.HandleDatabaseRecordChange(record);
+                    EtlLoader?.HandleDatabaseRecordChange(record);
+                    SubscriptionStorage?.HandleDatabaseRecordChange(record);
 
-                        OnDatabaseRecordChanged(record);
+                    OnDatabaseRecordChanged(record);
 
-                        LastDatabaseRecordChangeIndex = index;
+                    LastDatabaseRecordChangeIndex = index;
 
-                        if (_logger.IsInfoEnabled)
-                            _logger.Info($"Finish to process record {index} for {record.DatabaseName}.");
-                    }
-                    catch (Exception e)
-                    {
-                        if (_logger.IsInfoEnabled)
-                            _logger.Info($"Encounter an error while processing record {index} for {record.DatabaseName}.", e);
-                        throw;
-                    }
+                    if (_logger.IsInfoEnabled)
+                        _logger.Info($"Finished to process record with index `{index}` for {record.DatabaseName}.");
                 }
-                finally
+                catch (Exception e)
                 {
-                    if (taken)
-                        Monitor.Exit(_clusterLocker);
+                    if (_logger.IsInfoEnabled)
+                        _logger.Info($"Encountered an error while processing record with index `{index}` for {record.DatabaseName}.", e);
+                    throw;
                 }
             }
+            finally
+            {
+                Monitor.Exit(_clusterLocker);
+            }
+        }
+
+        private void TryTakeClusterLock()
+        {
+            var lockTaken = false;
+            while (lockTaken == false)
+            {
+                DatabaseShutdown.ThrowIfCancellationRequested();
+
+                Monitor.TryEnter(_clusterLocker, TimeSpan.FromSeconds(5), ref lockTaken);
```

Comment:
We must return if the lock was taken or not, and release it only if it was taken.

Reply:
it is always taken unless there is a database shutdown

Reply:
You can take the `_clusterLocker` in `NotifyFeaturesAboutXXX` methods as well and they can compete with each other..

Reply:
This is how it used to work. I just unified both usages.